runtime/src/accounts.rs

```diff
@@ -320,36 +332,88 @@ impl Accounts {
     /// Slow because lock is held for 1 operation instead of many
     pub fn store_slow(&self, fork: Fork, pubkey: &Pubkey, account: &Account) {
         let mut accounts = HashMap::new();
-        accounts.insert(pubkey, (account, 0));
+        accounts.insert(pubkey, account);
         self.accounts_db.store(fork, &accounts);
     }
 
     fn lock_account(
         locks: &mut HashSet<Pubkey>,
-        keys: &[&Pubkey],
+        credit_only_locks: &Arc<RwLock<HashMap<Pubkey, CreditOnlyLock>>>,
+        message: &Message,
         error_counters: &mut ErrorCounters,
     ) -> Result<()> {
-        // Copy all the accounts
-        for k in keys {
+        let (credit_debit_keys, credit_only_keys) = message.get_account_keys_by_lock_type();
+
+        for k in credit_debit_keys.iter() {
+            let credit_only_locks = credit_only_locks.read().unwrap();
+            if locks.contains(k)
+                || (credit_only_locks.contains_key(k)
+                    && *credit_only_locks
+                        .get(&k)
+                        .unwrap()
+                        .lock_count
+                        .lock()
+                        .unwrap()
+                        > 0)
+            {
+                error_counters.account_in_use += 1;
+                debug!("Account in use: {:?}", k);
+                return Err(TransactionError::AccountInUse);
+            }
+        }
+        for k in credit_only_keys.iter() {
             if locks.contains(k) {
                 error_counters.account_in_use += 1;
                 debug!("Account in use: {:?}", k);
                 return Err(TransactionError::AccountInUse);
             }
         }
-        for k in keys {
-            locks.insert(**k);
+
+        for k in credit_debit_keys {
+            locks.insert(*k);
+        }
+        let mut credit_only_writes: Vec<&Pubkey> = vec![];
+        for k in credit_only_keys {
+            let credit_only_locks = credit_only_locks.read().unwrap();
+            if let Some(credit_only_lock) = credit_only_locks.get(&k) {
+                *credit_only_lock.lock_count.lock().unwrap() += 1;
+            } else {
+                credit_only_writes.push(k);
+            }
+        }
+
```

Comment:
is there race between building the credit_only_writes and actually executing them below?

Reply:
I don't think so. Currently, the credit-debit lock (Accounts::account_locks mutex) is held for the length of this method. Since all transactions must have a credit-debit fee payer, that means no two transactions can execute this method at the same time. Since this is the only place that CreditOnlyLock entries are created, if a CreditOnlyLock does not exist at this point, there is no way it can be created before the credit_only_writes loop below.
The CreditOnlyLock can be read and/or the lock_count decremented by other concurrent banking steps, but that should not be a problem, as long as the entry itself cannot subsequently be created (if it doesn't exist at this point) or removed (if it does).

Reply:
the rationale for making a list of write lock work is then to allow non-lock_account() dudes (like who?) to get at the read_lock?

Reply:
Yep, dudes like store_accounts() > collect_accounts(), which updates the credit AtomicU64 for a credit-only account